### PR TITLE
Reduce peak memory usage of TensileCreateLibrary

### PR DESCRIPTION
**Pull request by Serge45**
## Brief ##
This PR reduces the peak memory usage of `TensileCreateLibrary` via customized YAML loader and tweaks on logic loading flow.

## Implementation
 - [x] Customized YAML loader(with pyyaml's event mechanism) for improvement on speed and memory usage.
 - [x] Add faster path to get child of root element in YAML file.
 - [x] Use above to properly filter out mismatched archs durling logic listing.
 - [x] Allow `ParallelMap2` to return generator instead of list for overlapping logic reading and library merging.

### TensileCreateLibrary
Tested on building gfx942 with 64 workers:
#### Before
mprof_gfx942_org
Peak memory usage: 
```bash
mprofile_20240904190006.dat     108930.523 MiB
```
#### After
mprof_gfx942_0907
```bash
mprofile_20240906183806.dat     70725.227 MiB
```
Peak memory usage reduces to ~65% with ~9% build time increase. 

### For Building with All Archs
Here's the memory footprint for building all archs with 32 workers:
mprof_all_0907
peak memory usage:
```bash
mprofile_20240907151401.dat     71030.605 MiB
```
### Customized YAML Loader
Here's the data for loading `aldebaran/110CU/GridBased/aldebaran_Cijk_Alik_Bljk_HHS_BH_Bias_AH_SAV.yaml` logic in repo:
| |file size|peak memory|time(sec)
|-|----|------|--|
|Customized|14 MiB|111 MiB|5.06|
|pyyaml built-in|14 MiB|1039 MiB|9.65|

### CODE REVIEW

**Review by KKyang — DISMISSED** [review]

File: tensilelite/requirements.txt [hunk 1]
```diff
@@ -1,7 +1,7 @@
 dataclasses; python_version == '3.6'
 pyyaml
 msgpack
-joblib
+joblib>=1.4.0
```
Comment (davidd-amd) [inline tensilelite/requirements.txt]: SLES has joblib 1.1.1.

File: tensilelite/Tensile/TensileCreateLibrary.py [hunk 2]
```diff
@@ -831,7 +831,7 @@ def buildObjectFileNames(kernelWriterAssembly, kernels, kernelHelperObjs):
   sourceLibFiles = []
   asmLibFiles = []
 
-  asmKernels = list([k for k in kernels if k['KernelLanguage'] == 'Assembly'])
+  asmKernels = (k for k in kernels if k['KernelLanguage'] == 'Assembly')
```
Comment (davidd-amd) [inline tensilelite/Tensile/TensileCreateLibrary.py:834]: Nice!

File: tensilelite/requirements.txt (hunk 1 above)
Comment (Serge45) [reply tensilelite/requirements.txt]: Hmm thanks for the reminding! originally I implemented with `concurrent` module, and then switched to `joblib.Parallel` for simplicity, maybe we need such fallback version if `joblib` >= 1.4 not available.

**Review by KKyang — DISMISSED** [review]

**Review by jichangjichang — DISMISSED** [review]
lgtm

**Comment by davidd-amd** [comment]
@Serge45 @jichangjichang do we know why these tests are failing? Is there a plan to resolve the failures soon?

**Review by KKyang — DISMISSED** [review]

**Review by KKyang — APPROVED** [review]

**Review by hcman2 — APPROVED** [review]

**Review by jichangjichang — APPROVED** [review]
lgtm

**Pull request merged**